From e3451096f00cbdbd7fa5909f77317bc0936c443e Mon Sep 17 00:00:00 2001
From: Ranbir Singh <Ranbir.Singh3@Dell.com>
Date: Tue, 24 Jan 2023 12:27:56 +0530
Subject: [PATCH 1/2] SecurityPkg/.../SecureBootConfigDxe: Fix FORWARD_NULL
 Coverity issues

Full pathname - SecurityPkg/VariableAuthenticated/SecureBootConfigDxe

The function UpdateDeletePage makes a jump to label ON_EXIT if
StartOpCodeHandle is found to be NULL at its beginning. At label
ON_EXIT, call is made to "HiiUpdateForm" unconditionally. In this
scenario, passing null pointer "StartOpCodeHandle" to "HiiUpdateForm",
will lead to CRASH which dereferences it.

Similar issue exists in functions LoadSignatureList and LoadSignatureData.

Add an appropriate NULL pointer check before calling "HiiUpdateForm".

Cc: Min Xu <min.m.xu@intel.com>
REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4266
Signed-off-by: Ranbir Singh <Ranbir.Singh3@Dell.com>
---
 .../SecureBootConfigImpl.c                    | 63 ++++++++++---------
 1 file changed, 34 insertions(+), 29 deletions(-)

diff --git a/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c b/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c
index 4299a6b5e56d..86447998d735 100644
--- a/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c
+++ b/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c
@@ -2639,15 +2639,14 @@ UpdateDeletePage (
   }
 
 ON_EXIT:
-  HiiUpdateForm (
-    PrivateData->HiiHandle,
-    &gSecureBootConfigFormSetGuid,
-    FormId,
-    StartOpCodeHandle,
-    EndOpCodeHandle
-    );
-
   if (StartOpCodeHandle != NULL) {
+    HiiUpdateForm (
+      PrivateData->HiiHandle,
+      &gSecureBootConfigFormSetGuid,
+      FormId,
+      StartOpCodeHandle,
+      EndOpCodeHandle
+      );
     HiiFreeOpCodeHandle (StartOpCodeHandle);
   }
 
@@ -3807,21 +3806,25 @@ LoadSignatureList (
   }
 
 ON_EXIT:
-  HiiUpdateForm (
-    PrivateData->HiiHandle,
-    &gSecureBootConfigFormSetGuid,
-    FormId,
-    StartOpCodeHandle,
-    EndOpCodeHandle
-    );
+  if (StartOpCodeHandle != NULL) {
+    HiiUpdateForm (
+      PrivateData->HiiHandle,
+      &gSecureBootConfigFormSetGuid,
+      FormId,
+      StartOpCodeHandle,
+      EndOpCodeHandle
+      );
+  }
 
-  HiiUpdateForm (
-    PrivateData->HiiHandle,
-    &gSecureBootConfigFormSetGuid,
-    FormId,
-    StartGotoHandle,
-    EndGotoHandle
-    );
+  if (StartGotoHandle != NULL) {
+    HiiUpdateForm (
+      PrivateData->HiiHandle,
+      &gSecureBootConfigFormSetGuid,
+      FormId,
+      StartGotoHandle,
+      EndGotoHandle
+      );
+  }
 
   SECUREBOOT_FREE_NON_OPCODE (StartOpCodeHandle);
   SECUREBOOT_FREE_NON_OPCODE (EndOpCodeHandle);
@@ -4293,13 +4296,15 @@ LoadSignatureData (
   //
   PrivateData->CheckArray = AllocateZeroPool (SIGNATURE_DATA_COUNTS (ListWalker) * sizeof (BOOLEAN));
 ON_EXIT:
-  HiiUpdateForm (
-    PrivateData->HiiHandle,
-    &gSecureBootConfigFormSetGuid,
-    FormId,
-    StartOpCodeHandle,
-    EndOpCodeHandle
-    );
+  if (StartOpCodeHandle != NULL) {
+    HiiUpdateForm (
+      PrivateData->HiiHandle,
+      &gSecureBootConfigFormSetGuid,
+      FormId,
+      StartOpCodeHandle,
+      EndOpCodeHandle
+      );
+  }
 
   SECUREBOOT_FREE_NON_OPCODE (StartOpCodeHandle);
   SECUREBOOT_FREE_NON_OPCODE (EndOpCodeHandle);
-- 
2.36.1.windows.1

